master/internal/trial.go

```diff
@@ -1096,17 +1096,29 @@ func (t *trial) terminated(ctx *actor.Context) {
 		return
 	}
 
-	if w, err := t.sequencer.Workload(); err != nil {
-		if !t.replaying {
-			if err := markWorkloadErrored(t.db, w); err != nil {
-				ctx.Log().Error(err)
+	var w workload.Workload
+	var err error
+	e := workload.Errored
+	switch {
+	case !t.sequencer.UpToDate():
+		w, err = t.sequencer.Workload()
+		if err != nil && !t.replaying {
+			if wErr := markWorkloadErrored(t.db, w); wErr != nil {
+				ctx.Log().
+					WithError(wErr).
+					Error("failed to mark workload errored when terminated")
 			}
 		}
+	case t.sequencer.PrecloseCheckpointWorkload() != nil:
+		w = *t.sequencer.PrecloseCheckpointWorkload()
```

Comment:
Isn't `t.sequencer.PrecloseCheckpointWorkload()` stateless?  Or in other words, wouldn't this sometimes return a workload independent of whether or not it was the preclose checkpoint that failed?

Is it safer to mark that we are running a preclose checkpoint workload when we start running it, and then check that mark later?

Reply:
Yeah, it will return something anyways, but we check if there is a workload first. This is the same logic we use to find how out next workload (also in trial.go). I could abstract that out and use it in both places, but this is definitely correct and I feel less error prone than marking and checking later.

Reply:
This is just "give me the next workload I was supposed to do", same as it pulls it earlier.

Reply:
Makes sense to me